### DISCUSSION TITLE
Be able to specify TLS version

### DISCUSSION THREAD

**Discussion by gallargit** [dotnet/SqlClient#2444] [Ideas]
The current version of the Microsoft.Data.SqlClient package (5.2.0) is unable to connect to Sql Server 2005, whenever you try connect you get something like this:  `SSL Provider, error: 0 - An existing connection was forcibly closed by the remote host`

This is caused by the 1.2 version of the TLS library [(see explanation here)](https://stackoverflow.com/questions/60512153/provider-ssl-provider-error-0-an-existing-connection-was-forcibly-closed-by), it seems SQL Server 2005 and early 2008 versions cannot use TLS 1.2 at all. Older versions of the SqlClient package (up to 5.0.2) could connect instead.

It would be useful to provide a parameter that would allow you to choose older TLS versions to connect to those older servers.

**Comment by x509cert** [comment]
TLS support comes from the OS, SQL itself does not get involved. I wrote some blurb on this https://learn.microsoft.com/en-us/sql/relational-databases/security/networking/connect-with-tls-1-3?view=sql-server-ver16 

Also, SQL Server 2005 has been out of support for 8 years!

**Reply by gallargit** [reply]
Thanks for your answer, the question then is, why SqlClient 5.0 could connect to Sql2005 but 5.1+ can't? Maybe it did not use TLS but some other encryption algorithm?

I'm using this library+SMO to pull out data from old servers and moving it to new ones. I'd like to have the possibility of connecting to Sql2005 which was present before. Sure I know it's ancient ;)

**Comment by JRahnama** [comment]
@gallargit, it's likely that you won't be able to utilize the latest driver versions to establish a connection with SQL Server 2005. Specifically, any version past v3.0.0 is incompatible with SQL Server 2005. The capability to establish a connection in v5.0 could stem from its potential support for TLS 1.3. It's important to note that we didn't enforce this; rather, if the client supports TLS 1.3 and Encryption = Strict is utilized in the connection string, TLS 1.3 will be attempted. In the event that the server doesn't support TLS 1.3, it will revert to 1.2.

Also, while it's possible to establish a connection with v5.0, it's not assured to work in all scenarios. Failures may occur on a case-by-case basis.

Furthermore, it's imperative to acknowledge that we cannot provide support for an unsupported version of SQL Server on the driver.

**Comment by gallargit** [comment]
That's more or less what I was expecting, and it certainly makes sense. Thank you for all the information.
